File: src/api/connectors.ts
import { invoke } from "@tauri-apps/api/core";

import type { ConnectorDefinition } from "@/lib/bindings/ConnectorDefinition";
import type { CreateConnectorDefinitionInput } from "@/lib/bindings/CreateConnectorDefinitionInput";
import type { UpdateConnectorDefinitionInput } from "@/lib/bindings/UpdateConnectorDefinitionInput";

// ============================================================================
// Connectors
// ============================================================================

export const listConnectors = () =>
  invoke<ConnectorDefinition[]>("list_connectors");

export const getConnector = (id: string) =>
  invoke<ConnectorDefinition>("get_connector", { id });

export const createConnector = (input: CreateConnectorDefinitionInput) =>
  invoke<ConnectorDefinition>("create_connector", { input });

export const updateConnector = (
  id: string,
  input: UpdateConnectorDefinitionInput,
) => invoke<ConnectorDefinition>("update_connector", { id, input });

export const deleteConnector = (id: string) =>
  invoke<boolean>("delete_connector", { id });

File: src-tauri/src/commands/credentials/connectors.rs
use std::sync::Arc;
use tauri::State;

use crate::db::models::{
    ConnectorDefinition, CreateConnectorDefinitionInput, UpdateConnectorDefinitionInput,
};
use crate::db::repos::resources::connectors as repo;
use crate::engine::api_proxy::{invalidate_connector_cache, refresh_connector_keyword_snapshot};
use crate::error::AppError;
use crate::AppState;
use personas_macros::requires;

#[tauri::command]
pub fn list_connectors(
    state: State<'_, Arc<AppState>>,
) -> Result<Vec<ConnectorDefinition>, AppError> {
    // Public command — no IPC token required (read-only, needed at startup)
    repo::get_all(&state.db)
}

#[tauri::command]
pub fn get_connector(
    state: State<'_, Arc<AppState>>,
    id: String,
) -> Result<ConnectorDefinition, AppError> {
    repo::get_by_id(&state.db, &id)
}

#[tauri::command]
#[requires(privileged)]
pub fn create_connector(
    state: State<'_, Arc<AppState>>,
    input: CreateConnectorDefinitionInput,
) -> Result<ConnectorDefinition, AppError> {
    let result = repo::create(&state.db, input)?;
    invalidate_connector_cache();
    refresh_connector_keyword_snapshot(&state.db);
    Ok(result)
}

#[tauri::command]
#[requires(privileged)]
pub fn update_connector(
    state: State<'_, Arc<AppState>>,
    id: String,
    input: UpdateConnectorDefinitionInput,
) -> Result<ConnectorDefinition, AppError> {
    let result = repo::update(&state.db, &id, input)?;
    invalidate_connector_cache();
    refresh_connector_keyword_snapshot(&state.db);
    Ok(result)
}

#[tauri::command]
#[requires(privileged)]
pub fn delete_connector(state: State<'_, Arc<AppState>>, id: String) -> Result<bool, AppError> {
    let result = repo::delete(&state.db, &id)?;
    invalidate_connector_cache();
    refresh_connector_keyword_snapshot(&state.db);
    Ok(result)
}
